cmd/libsnap-confine-private/snap-test.c

```diff
@@ -214,6 +231,82 @@ static void test_sc_snap_name_validate__respects_error_protocol(void)
 	    ("snap name must use lower case letters, digits or dashes\n");
 }
 
+static void test_sc_instance_name_validate(void)
+{
+	struct sc_error *err = NULL;
+
+	sc_instance_name_validate("hello-world", &err);
+	g_assert_null(err);
+	sc_instance_name_validate("hello-world_foo", &err);
+	g_assert_null(err);
+
+	// just the separator
+	sc_instance_name_validate("_", &err);
+	g_assert_nonnull(err);
+	g_assert_true(sc_error_match
+		      (err, SC_SNAP_DOMAIN, SC_SNAP_INVALID_INSTANCE_NAME));
+	g_assert_cmpstr(sc_error_msg(err), ==,
+			"snap name must contain at least one letter");
+	sc_error_free(err);
+
+	// just name, with separator, missing instance key
+	sc_instance_name_validate("hello-world_", &err);
+	g_assert_nonnull(err);
+	g_assert_true(sc_error_match
+		      (err, SC_SNAP_DOMAIN, SC_SNAP_INVALID_INSTANCE_KEY));
+	g_assert_cmpstr(sc_error_msg(err), ==,
+			"instance key must contain at least one letter or digit");
+	sc_error_free(err);
+
+	// only separator and instance key, missing name
+	sc_instance_name_validate("_bar", &err);
+	g_assert_nonnull(err);
+	g_assert_true(sc_error_match
+		      (err, SC_SNAP_DOMAIN, SC_SNAP_INVALID_INSTANCE_NAME));
+	g_assert_cmpstr(sc_error_msg(err), ==,
+			"snap name must contain at least one letter");
+	sc_error_free(err);
+
+	sc_instance_name_validate("", &err);
+	g_assert_nonnull(err);
+	g_assert_true(sc_error_match
+		      (err, SC_SNAP_DOMAIN, SC_SNAP_INVALID_NAME));
+	g_assert_cmpstr(sc_error_msg(err), ==,
+			"snap name must contain at least one letter");
+	sc_error_free(err);
+
+	const char *valid_names[] = {
+		"a", "aa", "aaa", "aaaa",
+		"a_a", "aa_1", "a_123", "a_0123456789",
+	};
+	for (size_t i = 0; i < sizeof valid_names / sizeof *valid_names; ++i) {
+		g_test_message("checking valid instance name: %s",
+			       valid_names[i]);
+		sc_instance_name_validate(valid_names[i], &err);
+		g_assert_null(err);
+	}
+	const char *invalid_names[] = {
+		// only letters and digits in the instance key
+		"a_--23))", "a_ ", "a_091234#", "a_123_456",
+		// up to 10 characters for the instance key
+		"a_01234567891", "a_0123456789123",
+		// snap name must not be more than 40 characters, regardless of instance
+		// key
+		"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx_foobar",
+		"xxxxxxxxxxxxxxxxxxxx-xxxxxxxxxxxxxxxxxxxx_foobar",
```

Comment:
This would be slightly easier to read if you used `01234567890123456789012345678901234567890_foobar`.